doc/design/error_handling_requirements.md

```diff
@@ -0,0 +1,228 @@
+
+### Recoverable Error Handling
+
+A recoverable error handling strategy allows the system to continue operation after the handling code
+was executed successfully. Exceptions are an example of such a mechanism but also any handler code
+that does not result in termination. If the handling code itself fails (e.g. insufficient resources)
+no recovery is possible and the system terminates.
+
+### Non-recoverable Error Handling
+
+A non-recoverable error handling strategy will always call terminate after the handling code was executed.
+Handler code may be used to initiate a graceful shutdown. An example are (strict) asserts.
+
+## Requirements
+
+In the following we list the requirements of the Error Handling (EH) strategy. The term platform
+refers to the usage environment of the error handler. There can be various platforms,
+e.g. for regular operations, testing, different OS etc.
+
+### Functional
+
+1. The EH shall support error codes.
+    - configurable on a platform basis
+    - errors from different platforms shall not be usable (checked at compile time)
+    - uniqueness is not enforced (can be done by the layer above if desired)
+    - error codes shall be optional (if none is used, it is mapped to some default error code)
+1. The EH shall support multiple error levels.
+    - the individual error levels are defined by the platform
+    - there has to be at least a FATAL error level (most severe error level)
+    - the error levels shall be ordered from least severe to most severe
+    - if the code requires an ERROR level that is not supported by the platform it shall not compile
+1. The EH shall support a non-recoverable error handling strategy in case of FATAL errors.
+    - non-recoverable errors shall call terminate after optional handling and logging code has been executed
+1. The EH shall support a recoverable error handling strategies for non-FATAL errors.
+    - recoverable error handling shall support execution of sufficiently general user actons (callbacks)
+1. The detection mechanism of the EH shall allow logging the error.
+    - user defined-logging shall be supported
+    - there shall be a default implementaton for logging
+1. The EH shall allow to trace the error to its source location (file, function, line).
+1. The EH shall be disableable for non-FATAL errors.
+    - disabling shall be possible up to some level at compile time; this level shall be non-FATAL
+    - can be used if there is sufficient trust in the system to disable e.g. warning checks
+    - by default all levels defined by the platform are enabled
+    - disabling shall cause the affected error logic in the code to cause no runtime overhead
+
+### Technical Constraints
+
+1. The EH shall be uniquely defined for each platform.
+    - only one error handler can be active for a platform at any time
+    - the EH shall not be changed while the system runs
+    - the EH must be configurable at runtime in the init phase
```

Comment:
Here I would explicitly not handle it like the logger. I would state that the EH must be only configurable at compile time or link time. Otherwise what happens when you reconfigure the error handler at runtime? You could forbid that but than you have to do error handling for the error handler.
You could allow it but then you open the door to a lot of misuse cases. What happens when multiple instance or change the error handler or an error occurrs concurrently while the error handler is changed.

Reply:
It is fine with me to configure it at compile time only. Makes stuff even easier.